File: python/PStest/2021_k/2.py
import time

prev_time = time.time_ns()

from itertools import combinations

def solution(orders, course):
    answer = []

    dict_list = [{} for _ in range(11)]

    for order in orders:
        part_order = []
        for i in range(1,len(order)+1):
            part_order.extend(combinations(order,i))
        # print(part_order)

        for part in part_order:
            # part_str = ''.join(set(part))
            part_str = ''.join(sorted(part))
            print(part_str)
            if part_str not in dict_list[len(part_str)]:
                dict_list[len(part_str)][part_str] = 0
            dict_list[len(part_str)][part_str] += 1
        # print(dict_list)

    for cour in course:
        sorted_order = sorted(dict_list[cour].items(), key=lambda x:x[1],reverse=True)
        # print('sorted_order', cour, sorted_order)

        if len(sorted_order) == 0:
            continue
        max_num = sorted_order[0][1]
        if max_num < 2:
            continue
        for order in sorted_order:
            # print(order)
            if max_num != order[1]:
                break
            answer.append(order[0])

    answer.sort()

    return answer


# orders = ["ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"]
# course = [2,3,4]

# orders = ["ABCDE", "AB", "CD", "ADE", "XYZ", "XYZ", "ACD"]
# course = [2,3,5]

orders = ["XYZ", "XWY", "WXA"]
course = [2,3,4]

print(solution(orders, course))


print('time', time.time_ns()-prev_time)

File: programmers/kakao 2021/2.cpp
//
// Created by Junho Kim on 2021/09/04.
//

#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

int N;
unordered_map<string, int> cmap;


void select(int depth, string order, string sel) {
    if (depth == sel.length()) {
        cmap[sel] += 1;
        return;
    }

    for (int i = 0; i < order.length(); ++i) {
        select(depth, order.substr(i+1), sel + order[i]);
    }
}


vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    cmap.clear();

    for (string &order : orders) {
        sort(order.begin(), order.end());
    }

    for (int crs : course) {
        for (string &order : orders) {
            select(crs, order, "");
        }

        int maxn = 0;
        for (auto it : cmap) {
            maxn = max(maxn, it.second);
        }

        for (auto it : cmap) {
            if (maxn >= 2 && it.second == maxn) {
                answer.push_back(it.first);
            }
        }
        cmap.clear();
    }

    sort(answer.begin(), answer.end());

    return answer;
}


#include <iostream>
using namespace std;

int main() {
//    vector<string> orders = {"ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"};
//    vector<int> course = {2,3,4};

    vector<string> orders = {"ABCDE", "AB", "CD", "ADE", "XYZ", "XYZ", "ACD"};
    vector<int> course = {2,3,5};
//
//    vector<string> orders = {"XYZ", "XWY", "WXA"};
//    vector<int> course = {2,3,4};

    solution(orders, course);


    return 0;
}